bundles/AdminBundle/Controller/Admin/TagsController.php

```diff
@@ -208,8 +210,10 @@ public function addTagToElementAction(Request $request)
 
         $tag = Tag::getById($tagId);
         if ($tag) {
+            $event = new TagEvent($tag, array('assignmentElementId' => $assginmentCId, 'assignmentElementType' => $assginmentCType));
+            \Pimcore::getEventDispatcher()->dispatch(TagEvents::PRE_ADD_TO_ELEMENT, $event);
```

Comment:
@simon-dcode I guess we should move the event dispatching into `Tag::addTagToElement()` so that it is also dispatched when used in custom scripts.

Reply:
If I dispatch this in `Tag::addTagToElement()`, how will I be able to differentiate between Tag assignments made on the Admin and made in the code somewhere else. 
Do you think I should create separated events for these situations then?

Reply:
Why should you differentiate between admin or API? 
This is also not done for other events as far as I know. 
One possibility would be to pass additional arguments to the event object, like `source`, which can be optionally set.

Reply:
It would be pointless to explain how my system works. The point is, I need to differentiate these two. But since the TagEvent object has the arguments, I'll handle this difference with that. Thanks for the suggestion.
The question is, can I send in the PR with this differentiating or is this 'too custom'?

Reply:
Good point, I've just checked if we have done this already in the past and unfortunately the answer is no, so I guess it would be inconsistent to do it now with tags 😟

So I guess it needs a workaround for your case. 
You can try the following in the event handler: 
```php
        $masterRequest = $requestStack->getMasterRequest();
        if($masterRequest) {
            $route = $masterRequest->get('_route');
            if($route === 'pimcore_admin_tags_addtagtoelement') {
                // do whatever you want
            }
        }
```

Not nice, but works :D

Reply:
you can also check for the context... if context is pimcore admin, you know the request comes from pimcore backend

Reply:
Could you give me a code example of what you're thinking of?

Reply:
https://github.com/coreshop/CoreShop/blob/b3f85464e9f9e77728ae5e4e704844b361bb3c8b/src/CoreShop/Component/Core/Locale/Context/PimcoreAdminLocaleContext.php#L75

Reply:
You don't need to pass a request object, the ContextResolver fetches the current request if null

Reply:
Thank you!